[Feature #289] Integrate PostgreSQL-based marker gene functionality

ke4 opened the pull request:
Create a new DAO to fetch gene expressions from the new gxa_marker_gene table in our PostgreSQL.

Every specific genes in each assay group are clustered. The new table will contains the priority for each gene in the marker_gene_rank column. In each assay group the data should be ordered by that column.

The assay group IDs (assay field in the DB table or with other name samples in the given experiment) are passed (already calculated beforehand) into this service and we use them as a query filter. They are representing the column name in the heatmap table.

We should also select the top X gene for every assay groups based on the number of assay groups divided by 50 and round down to its integer value.

For example:

number of assay groups = 18 -> 50 % 18 = 2
In this case we are showing 36 rows (2 * 18)
number of assay groups: 9 -> 50 % 9 = 5
In this case we are showing 45 rows (5 * 9)
Note: Later on this value can come from a slider from the UI.
The json output should be exactly same format as the previous one, to prevent further changes in the frontend side.

Changes:
- Add `MarkerGeneDao` to fetch marker gene profiles and counts from PostgreSQL when looking for `Most Specific` genes. - Updated `BaselineExperimentProfilesService` to conditionally use the new DAO for marker genes or the existing Solr-based implementations.
- Included unit tests for `MarkerGeneDao` and updated service tests for comprehensive coverage.
- Modified some related services / tests

ke4 committed:
Integrate PostgreSQL-based marker gene functionality

Add `MarkerGeneDao` to fetch marker gene profiles and counts from PostgreSQL when `specific` flag in preferences is true. Updated `BaselineExperimentProfilesService` to conditionally use the new DAO for marker genes or the existing Solr-based implementations. Included unit tests for `MarkerGeneDao` and updated service tests for comprehensive coverage.

ke4 committed:
Add gxa_marker_gene fixture scripts for integration tests

Introduced new database fixture and cleanup scripts for the `gxa_marker_gene` table to support integration testing. Updated relevant test classes to populate and clean this table during test setup and teardown phases. These changes enable testing scenarios that involve marker gene data.

ke4 committed:
Refactor `ExpressionUnit` to support database-specific values

Update `ExpressionUnit` enums to include a `getDatabaseValue()` method for consistent database representation. Adjust related service and DAO logic to use this method, ensuring accurate parameter handling. Enhance test coverage to validate new functionality.

ke4 committed:
Refactor snd clean MarkerGeneDao for clearer SQL and modular structure

Extracted reusable SQL fragments and modularized query-building logic for better readability and maintainability. Added helper methods to streamline SQL execution, parameter building, and results processing. Introduced stricter null constraints for method parameters to improve type safety.

ke4 committed:
More code cleaning and add javadoc comments to enhance code documentation

Detailed javadoc comments were added to DAO and service classes to explain their responsibilities, methods, and parameters. This improves code readability and helps developers understand the purpose and usage of each component more effectively.

ke4 committed:
Refactor BaselineExperimentProfilesServiceTest setup and mocks

Streamlined the test setup by introducing helper methods for initialization. Renamed mock variable names for clarity and updated to reflect switch from MarkerGeneDao to PostgresDao usage. Improved readability and maintained consistency across test cases.

ke4 committed:
Update expression unit to lowercase 'tpms' in test fixture

Replaced all instances of 'TPM' with 'tpms' in the GXA marker gene fixture to ensure consistency with expected data formats. This change aligns the test data with the application's case sensitivity requirements.

ke4 committed:
Get assay iD from column headers by assay name

Added logic to match assay names with assay groups using column headers, along with supporting utility methods for safer JSON handling and improved logging for unmatched cases.

ke4 committed:
Add mockColumnHeaders to align test cases in MarkerGeneDaoTest and BaselineExperimentProfilesServiceTest for the updated code

This update introduces mockColumnHeaders to ensure proper handling of assay group metadata in test cases.

ke4 committed:
Fix tests to use `getDatabaseValue()` for expression units to get the correct value from DB

Replaced hardcoded strings ("TPM", "FPKM") with `ExpressionUnit.Absolute.Rna.getDatabaseValue()` in test cases. This ensures consistency with database values and reduces potential errors from string mismatches.

ke4 committed:
Update project version to 37.2.0

Bump the version from 37.1.1 to 37.2.0 in the Gradle build script. This update likely includes changes or improvements aligning with the new version.

ke4 committed:
Refactor marker gene SQL queries that match the modified requirements

Added filtering by assay names and optional gene ids passed in by params.
Introduced a constant for maximum marker genes and adjusted logic accordingly.

ke4 committed:
Update gene query logic to support gene IDs or names - WIP

Refactored SQL queries and parameters to handle gene filtering by both gene IDs and gene names. Introduced additional checks for empty gene queries and adjusted query construction accordingly. Updated expression unit enums to include lowercase variants for consistent handling.

ke4 committed:
Refactor marker gene profile fetching logic

Consolidated logic for fetching marker gene profiles by replacing `fetchSpecificGeneProfiles` with `fetchMarkerGeneProfiles`. Introduced helper methods for cleaner query parameter handling and improved readability. Updated affected tests to align with the refactored method signatures.

ke4 committed:
Refactor MarkerGeneDaoTest to align to the changed code under test

Simplified test methods by removing redundant comments, consolidating common logic, and improving naming for better clarity. Introduced helper methods and reused constants to reduce duplication and ensure consistency across tests.

ke4 committed:
Set gene query in request preferences for baseline tests

Added a default SemanticQuery for gene queries in `RnaSeqBaselineRequestPreferences` initialization. This ensures consistent handling of gene-related queries during baseline test setups.

ke4 committed:
Refine gene profile fetching logic and Upgrade dependencies

Simplified logic in `BaselineExperimentProfilesService` by removing specific gene search handling.

ke4 committed:
Ensure minimum marker gene rank limit

Add a safeguard in MarkerGeneDao to set a minimum marker gene rank limit of 1 when the calculated value is less than 1.

ke4 committed:
Order marker genes by assay names and expression level

Replaced HashMap with LinkedHashMap to preserve insertion order. Updated SQL query to include ordering by assay names and descending expression levels. Adjusted query parameters to reflect the new ordering logic.

ke4 committed:
Filter marker genes without a rank

Enhance SQL query in `MarkerGeneDao` to exclude entries with null `marker_gene_rank` for improved data accuracy.

ke4 committed:
Refactor MarkerGeneDao to use assay group IDs consistently

Replaced assay names with assay group IDs for improved consistency and clarity in query logic and data mapping. Updated SQL query, helper methods, and associated logic to reflect this change. Removed unused code and redundant logging for better maintainability.

ke4 committed:
Remove unused parameter from fetchMarkerGeneProfiles calls

Eliminated the redundant "factorValue"/"name" parameter in fetchMarkerGeneProfiles method calls across the codebase. Updated corresponding test cases to align with the method signature changes for consistency and accuracy.

ke4 committed:
Refactor the SQL in MarkerGeneDao and use parameter substitution

Updated SQL queries to utilize `StringSubstitutor` for dynamic parameter substitution, improving readability and maintainability. Replaced positional parameters with named placeholders and refactored query parameter handling to generate a map of values for substitution. Removed redundant parameters in `executeQuery` as a substitution simplifies query preparation.

ke4 committed:
Update RNA expression units from singular to plural

Changed RNA expression unit enums from "fpkm" to "fpkms" and "tpm" to "tpms" to reflect pluralization. This ensures consistency and clarity in naming conventions.

ke4 committed:
Update test to use `assay_group_id` instead of `assay names`

Replaced references to `assay names` with `assay_group_id` for consistency with the updated data structure. Adjusted related test logic and method calls accordingly to reflect this change.

ke4 committed:
Refactor marker gene query ordering logic.

Replaced PostgreSQL's `ARRAY_POSITION` function with Java-based sorting to improve flexibility and maintainability. Updated the SQL query to remove `ARRAY_POSITION` and added a utility method to sort results by assay order in Java. This ensures consistent ordering of marker genes based on assay group IDs.

ke4 committed:
Add assay_id to marker gene table and update test constants

Updated the SQL schema to include a new `assay_id` column in the marker gene table and adjusted the test data accordingly. Refactored test constants and methods in `BaselineExperimentProfilesServiceTest` to align with the updated schema.

sandsebi commented on app/src/main/java/uk/ac/ebi/atlas/experimentpage/baseline/profiles/MarkerGeneDao.java:
```diff
@@ -0,0 +1,314 @@
+    private static final int MAX_NUMBER_OF_MARKER_GENES = 50;
+    private final JdbcTemplate jdbcTemplate;
+
+    private static final String FETCH_MARKER_GENES =
```
Any value passed into preferences or columnHeaders from an external source could be manipulated to execute malicious database commands.
Recommendation: Do not use StringSubstitutor for SQL queries. Use a parameterized query with JdbcTemplate's built-in methods. Instead of ${variable}, use ? placeholders and pass the values in an array or a map.
Eg:

private static final String FETCH_MARKER_GENES_SQL =
        "SELECT gene_id, gene_name, assay_id, expression_level, marker_gene_rank " +
        "FROM gxa_marker_gene " +
        "WHERE experiment_accession = :experiment_accession " +
        "AND assay_id IN (:assay_ids) " +
        "AND gene_id IN ( " +
            "SELECT DISTINCT gene_id " +
            "FROM gxa_marker_gene " +
            "WHERE experiment_accession = :experiment_accession " +
            "AND assay_id IN (:assay_ids) " +
            "AND marker_gene_rank <= :marker_gene_rank " +
            "AND expression_unit = :expression_unit " +
            "AND expression_level >= :expression_level) " +
        "AND expression_unit = :expression_unit " +
        "AND expression_level >= :expression_level " +
        "ORDER BY marker_gene_rank IS NULL, " +
        "marker_gene_rank, " +
        "expression_level DESC";
private final NamedParameterJdbcTemplate namedParameterJdbcTemplate;

    public MarkerGeneDao(JdbcTemplate jdbcTemplate) {
        // Wrap the existing JdbcTemplate in a NamedParameterJdbcTemplate.
        this.namedParameterJdbcTemplate = new NamedParameterJdbcTemplate(jdbcTemplate);
    }
    
    // ... (unchanged fetchMarkerGeneProfiles method signature)
    public GeneProfilesList<BaselineProfile> fetchMarkerGeneProfiles(
        @NotNull String experimentAccession,
        @NotNull List<AssayGroup> assayGroups,
        @NotNull BaselineRequestPreferences<?> preferences,
        @NotNull JsonArray columnHeaders) {

        var markerGeneRankLimit = (double) (MAX_NUMBER_OF_MARKER_GENES / columnHeaders.size());
        if (markerGeneRankLimit < 1) {
            markerGeneRankLimit = 1;
        }

        final List<String> assayGroupIDs = getAssayGroupIDs(columnHeaders);
        
        // Use a MapSqlParameterSource to securely bind parameters.
        var params = new MapSqlParameterSource()
            .addValue("experiment_accession", experimentAccession)
            .addValue("assay_ids", assayGroupIDs)
            .addValue("marker_gene_rank", markerGeneRankLimit)
            .addValue("expression_unit", preferences.getUnit().getDatabaseValue())
            .addValue("expression_level", preferences.getCutoff());

        var results = namedParameterJdbcTemplate.queryForList(FETCH_MARKER_GENES_SQL, params);
        
        // ... (rest of the method remains the same)
    }

ke4 replied:
TBH I did that one before, but it did not look clean for me and refactored with this one.
But I accept you concern. I will try another way to do it.
Thanks

ke4 replied:
OK, I reread your comment now and I think I understand what you are suggesting, although you wrote a bit confusing sentence: `Instead of ${variable}, use ? placeholders and pass the values in an array or a map.` but I think you meant to use named parameters instead of String replacement and use the `NamedParameterJdbcTemplate `.
I am going to do that.
Thanks

sandsebi commented on app/src/main/java/uk/ac/ebi/atlas/experimentpage/baseline/profiles/MarkerGeneDao.java:
```diff
@@ -0,0 +1,314 @@
+     * @param results The database query results to sort
+     * @param assayGroupIDs The ordered list of assay group IDs
+     */
+    private void sortResultsByAssayOrder(List<Map<String, Object>> results, List<String> assayGroupIDs) {
```
I thought sorting the data in the database and returning it is a more efficient approach. Do you mind explaining why this manual approach is being used?

ke4 replied:
I agree and I did that before, but the query was too PostgreSQL specific. We use H2 for executing the unit tests and that can't execute the PostgreSQL specific part. I added that comment for this commit: [89eb8ef](https://github.com/ebi-gene-expression-group/atlas-web-bulk/pull/296/commits/89eb8eff4d84909dff11c5ba70f41fbc445c43f4)
I am not sure there is a way to do this specific ordering without using some DB specific function.

sandsebi replied:
Okay 👍

sandsebi reviewed (commented):
Added my comments.

ke4 committed:
Refactor MarkerGeneDao to use NamedParameterJdbcTemplate

Replaced JdbcTemplate with NamedParameterJdbcTemplate for better SQL parameter handling and readability. Updated related methods, queries, and tests to accommodate the change. This ensures improved maintainability and reduces the risk of SQL injection.

sandsebi reviewed (approved):
Looks good 👍
